### PR TITLE
Buttons and sensor

### PR DESCRIPTION
**Pull request by JayWai99**
Lemme know if there is anything I need to add to my part of the code.

### CODE REVIEW

File: include/power_button.h [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#ifndef POWER_BUTTON_H
+#define POWER_BUTTON_H
+
+#include <stdbool.h>
+
+void setup_power_button(void);
+
+bool is_pressed_power_button(void);
+bool is_pressed_power_button_debounced(void);
```
Comment (MitchellUTS) [inline include/power_button.h]: Why have you changed the functions available for device?

File: include/power_button.h [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+bool power_button_state(bool button_state);
+
+#endif // LED_BUTTON_H
```
Comment (MitchellUTS) [inline include/power_button.h]: I don't know why you decided to rename everything, when this button doesn't control the power. It just controls the LED, but at least please make sure your fix the comments when renaming.

File: include/ultrasonic.h [hunk 3]
```diff
@@ -4,7 +4,7 @@
 #include <stdint.h>
 
 void setup_ultrasonic(void);
-
+bool is_ultrasonic_echo_high(void);
```
Comment (MitchellUTS) [inline include/ultrasonic.h]: Once again, justify why you need this as existing specification should be sufficient.
I'm happy to make changes but I don't see why this change is needed.

File: include/volume_button.h [hunk 4]
```diff
@@ -6,5 +6,6 @@
 void setup_volume_button(void);
 
 bool is_pressed_volume_button(void);
+bool is_volume_button_pressed_debounced(void);
```
Comment (MitchellUTS) [inline include/volume_button.h:8]: Once again, random new API changes without a clear reason.

File: src/power_button.cpp [hunk 5]
```diff
@@ -0,0 +1,40 @@
+#include <avr/io.h>
+#include <util/delay.h>
+#include "power_button.h"
+#include "bit.h"
+
+#define POWER_BUTTON_PIN PINB3
+#define DEBOUNCE_DELAY 100 // CHANGE THIS
```
Comment (MitchellUTS) [inline src/power_button.cpp]: Please remove the 'CHANGE THIS' comments if you have changed them.

File: src/power_button.cpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+bool is_pressed_power_button_debounced(void) {
+    if (is_pressed_power_button()) {
+        _delay_us(DEBOUNCE_DELAY);
+        return is_pressed_power_button();
+    }
+    return false;
+}
+
+// bool power_button_state(bool button_state){
```
Comment (MitchellUTS) [inline src/power_button.cpp]: Is there any reason you left this comment block in?

File: src/power_button.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    BIT_SET(PORTB, POWER_BUTTON_PIN);
+
+    // Set the push button as input
+    BIT_CLEAR(DDRB, POWER_BUTTON_PIN);
+}
+
+// This function should return true if the power button is pressed and false otherwise.
+// This function should be debounced. If the button status isn't changed from the last known value it should return immediately.
+// Otherwise, it should wait so the output can be debounced (verified) and then return the new button status.
+bool is_pressed_power_button(void) {
```
Comment (MitchellUTS) [inline src/power_button.cpp:26]: The comment indicate this should be debounce, but it doesn't appear to be. Can you resolve this.

File: src/volume_button.cpp [hunk 8]
```diff
@@ -1,13 +1,31 @@
-#define VOL_BUTTON_PIN 0 // CHANGE THIS
+#include <avr/io.h>
+#include <util/delay.h>
+#include "bit.h"
+
+#define VOL_BUTTON_PIN PINB4 // CHANGE THIS
```
Comment (MitchellUTS) [inline src/volume_button.cpp]: remove redundant comments

File: src/volume_button.cpp [hunk 9]
```diff
@@ -1,13 +1,31 @@
...
+    BIT_SET(PORTB, VOL_BUTTON_PIN);
 
+    // Set the push button as input
+    BIT_CLEAR(DDRB, VOL_BUTTON_PIN);
 }
 
 // This function should return true if the led button is pressed and false otherwise.
 // This function should be debounced. If the button status isn't changed from the last known value it should return immediately.
 // Otherwise, it should wait so the output can be debounced (verified) and then return the new button status.
 bool is_pressed_volume_button(void) {
```
Comment (MitchellUTS) [inline src/volume_button.cpp:25]: again, this needs to be debounced as per the comment

**Review by MitchellUTS — CHANGES_REQUESTED** [review]
@JayWai99 Can you read feedback, fix where appropriate and once tested push a new commit.

File: include/ultrasonic.h (hunk 3 above)
Comment (JayWai99) [reply include/ultrasonic.h]: I will remove it as you won't use it in main.cpp. That was my bad.

File: include/power_button.h (hunk 2 above)
Comment (JayWai99) [reply include/power_button.h]: If there is a different name that makes more sense such as led_and_buzzer_button and you are happy with that name then I am open to change. Otherwise, I am pretty sure this particular button handles both the led and buzzer. I also understood that I was stepping out of my boundaries a good amount of times when we had that conversation during the lab. Again, I am open to make changes if you want me to as the code integrator. I hope we are cool with that.

File: src/power_button.cpp (hunk 6 above)
Comment (JayWai99) [reply src/power_button.cpp]: Nope. Will remove that.

File: src/power_button.cpp (hunk 7 above)
Comment (JayWai99) [reply src/power_button.cpp:26]: Right. On it.
